contrib/masterfiles-stage/common.sh

```diff
@@ -12,146 +12,217 @@ if [ ! "$debug_mode" == "true" ]; then
-  #   - Fetch updates
-  #   - Stash any local changes
-  #   - Checkout the proper branch
-  #   - Hard reset to remove any changes
-  #   - Validate the staged policy
-  #   - If Staged policy has validated and the release IDs are different
-  #     synchronize to final location (should we sync anyway to make sure the
-  #     distribution point is good?)
-
-  if ! type "git" > /dev/null; then
-    error_exit "git not found on path: ${PATH}"
-  fi
+git_stage_refspec() {
+  # This function depends on git_setup_local_mirrored_repo
+  # having been run, such that the variable local_mirrored_repo
+  # contains the (local) path to a bare git repository.
+  #
+  # (A mirror repository is a special case of a bare repository;
+  # either will work for this function but a bare non-mirror
+  # repository will have edge cases that are mishandled.)
+  #
+  # This function accepts a single argument: refspec,
+  # which should be a git tagname, branch, or commit hash.
+  #
+  # This function stages the refspec to the STAGING_DIR
+  # from local_mirrored_repo
 
-  # If we have a git checkout ensure the origin is set properly, update it and
-  # make sure all modified, extra, or missing files are reset so that we have a
-  # clean checkout
-  if [ -d "${STAGING_DIR}/.git" ]; then
-      cd "${STAGING_DIR}" && git remote set-url origin "${GIT_URL}" && (
-          git fetch -q origin || error_exit "Failed: git fetch -q origin"
-          git stash -q  || error_exit "Failed: git stash -q"
-          git checkout -q "${GIT_TAG_OR_COMMIT}" || error_exit "Failed: git checkout -q ${GIT_TAG_OR_COMMIT}"
-          # git pull --rebase origin "${GIT_TAG_OR_COMMIT}" || error_exit "Failed: git pull --rebase origin ${GIT_TAG_OR_COMMIT}"
-	  # Not sure we would want to rebase changes either, we want a clean sync with the upstream
-          # The above line appears to be a mistake as it will ALWAYS fail if given a tagname or commit hash.
-          # It will succeed if given "tags/tagname" or if given a branch name, but never with a bare tagname or commit hash.
-          git reset -q --hard "${GIT_TAG_OR_COMMIT}" || error_exit "Failed: git reset -q --hard ${GIT_TAG_OR_COMMIT}"
-          git clean -f || error_exit "Failed: git clean -f"
-          git clean -fd || error_exit "Failed: git clean -fd"
-      ) || error_exit "Failed to stage '${GIT_TAG_OR_COMMIT}' in '${STAGING_DIR}' from '${GIT_URL}'" 
-
-  # If we don't have a git clone wipe the directory and create a new clone and
-  # ensure the proper branch/tag is checked out.
-  else
-      rm -rf "${STAGING_DIR}"/* "${STAGING_DIR}"/.??*
-      git clone --no-hardlinks "${GIT_URL}" "${STAGING_DIR}" && cd "${STAGING_DIR}" && git checkout "${GIT_TAG_OR_COMMIT}" || error_exit "Failed to stage '${GIT_TAG_OR_COMMIT}' from '${GIT_URL}'"
+  mkdir -p "${STAGING_DIR}" || error_exit "Failed: mkdir -p '$STAGING_DIR'"
+  git --git-dir="${local_mirrored_repo}" --work-tree="${STAGING_DIR}" checkout -q -f "$1" ||
```

Comment:
@nickanderson, the obvious (and very inelegant) workaround to allow cf_promises_release_id to be populated with the contents of `.git/HEAD` when `cf-promises -T` is run (something you brought up elsewhere), would be to just copy that single file from the `${local_mirrored_repo}` in this function.  :)

Aside from that, do you just want the VCS_TYPEs GIT and GIT_TAG_OR_COMMIT to use typical `git clone`s?  (It introduces a *lot* more edge cases and potential bugs that way; I don't think the removed lines 110-131 even handle all the edge cases, just some of them.  I'm not going to be using those VCS_TYPEs personally so I don't have any forceful opinion.)